### DISCUSSION TITLE
GetFractal does not return any values?

### DISCUSSION THREAD

**Discussion by preintel** [DaveSkender/Stock.Indicators#599] [Q&A] [answered]
Attempting to use the GetFractals() on my quotes (https://dotnet.stockindicators.dev/indicators/Fractal/#content)
however it seems to never return a fractal result for my quotes, I tried a few variations of calling it and different quote data but seems to always return null, unless I am missing something?

```
//IEnumerable<FractalResult> fr = quotes.GetFractal(5);
IEnumerable<FractalResult> fr = quotes.GetFractal((int)(quotes.Count / 2 - 1));

                    foreach (FractalResult r in fr)
                    {
                        if (r.Date != null && r.FractalBear != null)
                            FractalBearList.Add(new DateTimePoint(r.Date, (double)r.FractalBear));
                        else if (r.Date != null && r.FractalBull != null)
                            FractalBullList.Add(new DateTimePoint(r.Date, (double)r.FractalBull));
                    }

                    MessageBox.Show("Fractal counts: " + FractalBearList.Count + " , " + FractalBullList.Count);
```

**Comment by DaveSkender** [comment, marked as answer]
`quotes.Count / 2 - 1)` as the fractal `windowSize` is likely the problem in your example. By definition, you have a Williams' Fractal only when the evaluation point is the highest or lowest point within +/- `windowSize` bars.  Try `2` here and use at least 100 quotes and see if you get any results.

Also, keep in mind, we return the same number of elements as there are in this historical quotes.  Since there are relatively fewer identified Fractals, most of the result values will be `null`.

**Reply by preintel** [reply]
I tried small windowSizes of 2 and 3 it works to produce the points, now trying to figure out the logic of building support/resistance lines from these points or deciding if it is better using Pivots indicator to determine support/resistance?
Would end up using 3+ points of FractalBulls that are in proximity in y-value to each other to determine the support and 3+ FractalBears to calculate resistance line. Alternative would be to use the HH and LL of the Pivots indicator to do this... Not sure which one would be better suited for this.

**Reply by DaveSkender** [reply]
Not sure if you’ve seen it, I’ve been struggling with this too, but for a divergence calculation: #434.  I think in the end, I like [Donchian Channels](https://daveskender.github.io/Stock.Indicators/indicators/Donchian/#content) for H/L support and resistance.

**Reply by preintel** [reply]
I read over that post and calculating the divergence of oscillator and price charts is more complicated than establishing a general support/resistance line:

1) The biggest issue is that future price action does not need to correlate with divergence of a technical oscillator and current price, although this is seen at times for certain stocks this is not a rule but an assumption as future price can go in either direction or remain flat regardless of divergence of the oscillator. (However we will take this assumption at face value since we are coding with this goal in mind).

2) Current/past price patterns and technical oscillator patterns can hide inside larger patterns (ie: determining a divergence of the oscillator with price locally may not show the same divergence if you "zoom out" to look at the surrounding data points, a uptrend of an oscillator can present itself inside a overall oscillator downtrend and may be read as an uptrend by the program). For this would need multiple orders of calculations including determining the local trend pattern of the price and oscillator then considering prior data to determine if these trends are actually all encompassing or there is another encapsulating trend. At higher order calculations would benefit from heuristics to determine patterns (http://web.ist.utl.pt/~adriano.simoes/tese/referencias/Papers%20-%20Adriano/Bull%20Flag.pdf), 
however I suspect would also need to adopt deep learning to these solutions. Finding trends where none exists, or missing trends that exist always remains a problem.

3) The order of magnitude of the higher high and lower low may be important in certain cases, for example in the chart, the oscillator low at x ~485 is calculated as a low value based on the slightly higher value following it (although in this case it appears to be noise and this is followed by stock price going down and the oscillator reflecting that). Looking at the stock price chart one can argue the price is range bound/consolidating from x ~445 to ~485 followed by a decline and the oscillator follows the same downtrend with a delay. It  is also unclear the underlying volume of stock that is trading to produce those results which complicates correlating the oscillator with price action.
[https://user-images.githubusercontent.com/8432125/131228181-ff99eeb1-b444-44d8-8e91-1144cdb7e565.png](url)

4) Oscillators rely on variables which can produce different results depending on the variables inputted, for example the look back period for an RSI will create a different calculation for the chart complicating the issue of determining divergence patterns.

**Reply by DaveSkender** [reply]
Thanks for the info; adding some of these links to my reading list.
